docs/usage/service-worker.md

```diff
@@ -0,0 +1,417 @@
+        counter += 1
+        dict = data.to_py()
+        dict['count'] = counter
+        return dict
+    `,
+    { globals: namespace },
+  );
+
+  // assign the modify_data function from the Python context to a Javascript variable
+  modifyData = namespace.get("modify_data");
+  namespace.destroy();
+});
+
+self.addEventListener("fetch", (event) => {
+  if (event.request.url.endsWith("json")) {
+    if (!modifyData) {
+      // For this example, throw so it's clear that the worker isn't ready to modify responses
+      // This is because we don't want to return a response that isn't modified yet
+      // If your service worker would return the same response as a server (eg. it's just performing calculations closer to home)
+      // then you may want to let the event through without doing anything
+      event.respondWith(
+        Promise.reject("Python code isn't set up yet, try again in a bit"),
+      );
+    } else {
+      event.respondWith(
+        // We aren't using the async await syntax because event.respondWith needs to respond synchronously
+        // it can't be executing after an awaited promise within the fetch event handler, otherwise you'll get this
+        // Uncaught (in promise) DOMException: Failed to execute 'respondWith' on 'FetchEvent': The event has already been responded to
+        fetch(event.request)
+          .then((v) => v.json())
+          .then((originalData) => {
+            let proxy = modifyData(originalData);
+            let pyproxies = [];
+
+            // Because toJs gives us a Map, we transform it to a plain Javascript object before changing it to JSON
+            let result = JSON.stringify(
+              Object.fromEntries(
+                proxy.toJs({
+                  pyproxies,
+                }),
+              ),
+            );
+            // Craft the new JSON response
+            return new Response(result, {
+              headers: { "Content-Type": "application/json" },
+            });
+          }),
+      );
+    }
+  }
+});
+
+// Code below is for easy iteration during development, you may want to remove or modify in a prod environment:
+
+// Immediately become the active service worker once installed, so we don't have a stale service worker intercepting requests
+// You can remove this code and achieve a similar thing by enabling "Update on Reload" in devtools, if supported:
+// https://web.dev/service-worker-lifecycle/#update-on-reload
+self.addEventListener("install", function () {
+  self.skipWaiting();
+});
+
+// With this, we won't need to reload the page before the service worker can intercept fetch requests
+// https://developer.mozilla.org/en-US/docs/Web/API/Clients/claim#examples
+self.addEventListener("activate", function (event) {
+  event.waitUntil(self.clients.claim());
+});
+```
+
+### XMLHttpRequest polyfill
+
+This script should be imported into your service worker, to assign `XMLHttpRequest` in the service worker's global scope.
```

Comment:
Let's not put the polyfill in the docs. We can give people a url or something.

Reply:
I feel like people might have trouble finding a good polyfill that works out of the box from npm. Most haven't been updated in years and the one I've vendored here has a bug where it assigns the wrong variable and breaks when used outside of Node.js. I'll try looking for one though.

Reply:
In that case, I am okay with including this as a separate file `.js` in this folder and linking to it.

Reply:
The author of xhr-shim responded quickly to a PR and now we can just recommend that package instead!